tests/test_transform_dead_code.py

```diff
@@ -0,0 +1,52 @@
+# (C) Copyright 2018- ECMWF.
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+import pytest
+
+from conftest import available_frontends
+from loki import Subroutine, FindNodes, Conditional, Assignment
+from loki.transform import dead_code_elimination
+
+
+@pytest.mark.parametrize('frontend', available_frontends())
+def test_transform_dead_code_conditional(frontend):
+    """
+    Test correct elimination of unreachable conditional branches.
+    """
+    fcode = """
+subroutine test_dead_code_conditional(a, b, flag)
+  real(kind=8), intent(inout) :: a, b
+  logical, intent(in) :: flag
+
+  if (flag) then
+    if (1 == 6) then
+      a = a + b
+    else
+      b = b + 2.0
+    end if
+
+    if (2 == 2) then
+      b = b + a
```

Comment:
Would you mind adding a second test with conditionals with `elseif` branches that exercise the permutations of options which paths are to be eliminated (if, elseif, else, if&elseif, elseif&else)? I expect this to largely work as expected without further work thanks to the multi-body conditionals being converted to if/else blocks. But the issue discovered by Paddy makes me think we may lack an update of the corresponding `is_elseif` flag.

Reply:
Yes, I've added two small tests for nested conditionals - one to the base test that gets fully resolved, and one additional test for partially resolved nested conditionals. The bug you feared never materialised though. :wink:

Reply:
Yeah, because you're leaving out the case that the conditional remains but the elseif branch is eliminated. See below for a suggestion how to amend the test